### PR TITLE
RIA-4081 Reverse the order of share-a-case calls and logging

### PR DESCRIPTION
**Pull request by tarun-palisetty**
### JIRA link (if applicable) ###
[RIA-4081](https://tools.hmcts.net/jira/browse/RIA-4081)

### Change description ###
Reversed the share-a-case calls, and more logging

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/clients/ClientResponseErrorHandler.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import org.springframework.web.client.DefaultResponseErrorHandler;
+
+public class ClientResponseErrorHandler extends DefaultResponseErrorHandler  {
+
+    @Override
+    public void handleError(ClientHttpResponse response) throws IOException {
+
+        if (response.getStatusCode().is4xxClientError() || response.getStatusCode().is5xxServerError()) {
+
+            try(BufferedReader reader = new BufferedReader(new InputStreamReader(response.getBody()))) {
```
Comment (makowalski) [inline src/main/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/clients/ClientResponseErrorHandler.java:17]: please check if response.getBody() is not null

**Review by burale — APPROVED** [review]
LGTM

File: src/main/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/clients/ClientResponseErrorHandler.java (hunk 1 above)
Comment (tarun-palisetty) [reply src/main/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/clients/ClientResponseErrorHandler.java]: Null check added

File: src/main/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/clients/ClientResponseErrorHandler.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import java.net.URI;
+import java.util.stream.Collectors;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.client.ClientHttpResponse;
+import org.springframework.web.client.DefaultResponseErrorHandler;
+
+@Slf4j
+public class ClientResponseErrorHandler extends DefaultResponseErrorHandler  {
```
Comment (makowalski) [inline src/main/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/clients/ClientResponseErrorHandler.java:16]: if you could add `@Component` annotation then you have a way to inject dependency by constructor in `CcdCaseAssignment` class

File: src/main/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/clients/ClientResponseErrorHandler.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    public boolean hasError(ClientHttpResponse response) throws IOException {
+        HttpStatus httpStatus = response.getStatusCode();
+        return httpStatus.is4xxClientError() || httpStatus.is5xxServerError();
+    }
+
+    @Override
+    public void handleError(URI url, HttpMethod method, ClientHttpResponse response) throws IOException {
+
+        String responseAsString = toString(response.getBody());
+        log.error("URL: {}, HttpMethod: {}, ResponseBody: {}", url, method, responseAsString);
```
Comment (makowalski) [inline src/main/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/clients/ClientResponseErrorHandler.java:28]: do we see ccd id within this log? previous way could track the issue quite easily, you will get only ccd id in the snow tickets, worth to consider keeping old error logic handling

File: src/main/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/clients/CcdCaseAssignment.java [hunk 4]
```diff
@@ -155,25 +136,13 @@ public void applyNoc(
...
-                    aacUrl + applyNocAssignmentsApiPath,
-                    HttpMethod.POST,
-                    requestEntity,
-                    Object.class
-                );
-
-        } catch (RestClientResponseException e) {
-            throw new CcdDataIntegrationException(
-                "Couldn't apply noc AAC case assignment for case ["
-                + callback.getCaseDetails().getId()
```
Comment (makowalski) [inline src/main/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/clients/CcdCaseAssignment.java:171]: the old logging matches ccd id with the problem, I am not sure if the new logging supports this, can you check?

**Review by makowalski — CHANGES_REQUESTED** [review]
please clarify my comments

**Review by makowalski — APPROVED** [review]
LGTM

**Pull request merged**